library/init/category/state.lean

```diff
@@ -1,124 +1,147 @@
-
-instance (σ : Type u) : monad (state σ) :=
-{pure := @state_return σ, bind := @state_bind σ,
- id_map := begin
-   intros, funext,
-   simp [state_bind], cases x s,
-   apply rfl
- end,
- pure_bind := by intros; apply rfl,
- bind_assoc := begin
-   intros, funext,
-   simp [state_bind], cases x s,
-   apply rfl
- end}
-end
+  variables {σ : Type u} {m : Type u → Type v}
 
-namespace state
-@[inline] def read {σ : Type u} : state σ σ :=
-λ s, (s, s)
+  variable  [monad m]
+  variables {α β : Type u}
 
-@[inline] def write {σ : Type} : σ → state σ unit :=
-λ s' s, ((), s')
+  @[inline] protected def pure (a : α) : state_t σ m α :=
+  ⟨λ s, pure (a, s)⟩
 
-@[inline] def write' {σ : Type u} : σ → state σ punit :=
-λ s' s, (punit.star, s')
-end state
+  @[inline] protected def bind (x : state_t σ m α) (f : α → state_t σ m β) : state_t σ m β :=
+  ⟨λ s, do (a, s') ← x.run s,
+           (f a).run s'⟩
 
-def state_t (σ : Type u) (m : Type u → Type v) [monad m] (α : Type u) : Type (max u v) :=
-σ → m (α × σ)
+  instance : monad (state_t σ m) :=
+  { pure := @state_t.pure _ _ _, bind := @state_t.bind _ _ _ }
 
-section
-  variable  {σ : Type u}
-  variable  {m : Type u → Type v}
-  variable  [monad m]
-  variables {α β : Type u}
+  protected def orelse [alternative m] {α : Type u} (x₁ x₂ : state_t σ m α) : state_t σ m α :=
+  ⟨λ s, x₁.run s <|> x₂.run s⟩
+
+  protected def failure [alternative m] {α : Type u} : state_t σ m α :=
+  ⟨λ s, failure⟩
+
+  instance [alternative m] : alternative (state_t σ m) :=
+  { failure := @state_t.failure _ _ _ _,
+    orelse  := @state_t.orelse _ _ _ _ }
+
+  @[inline] protected def get : state_t σ m σ :=
+  ⟨λ s, pure (s, s)⟩
+
+  @[inline] protected def put : σ → state_t σ m punit :=
+  λ s', ⟨λ s, pure (punit.star, s')⟩
+
+  @[inline] protected def modify (f : σ → σ) : state_t σ m punit :=
+  ⟨λ s, pure (punit.star, f s)⟩
+
+  @[inline] protected def lift {α : Type u} (t : m α) : state_t σ m α :=
+  ⟨λ s, do a ← t, pure (a, s)⟩
 
-  def state_t_return (a : α) : state_t σ m α :=
-  λ s, show m (α × σ), from
-    return (a, s)
+  instance : has_monad_lift m (state_t σ m) :=
+  ⟨@state_t.lift σ m _⟩
 
-  def state_t_bind (act₁ : state_t σ m α) (act₂ : α → state_t σ m β) : state_t σ m β :=
-  λ s, show m (β × σ), from
-     do (a, new_s) ← act₁ s,
-        act₂ a new_s
+  @[inline] protected def monad_map {σ m m'} [monad m] [monad m'] {α} (f : Π {α}, m α → m' α) :
+    state_t σ m α → state_t σ m' α :=
+  λ x, ⟨λ st, f (x.run st)⟩
+
+  instance (σ m m') [monad m] [monad m'] : monad_functor m m' (state_t σ m) (state_t σ m') :=
+  ⟨@state_t.monad_map σ m m' _ _⟩
+
+  -- TODO(Sebastian): uses lenses as in https://hackage.haskell.org/package/lens-4.15.4/docs/Control-Lens-Zoom.html#t:Zoom ?
+  protected def zoom {σ σ' α : Type u} {m : Type u → Type v} [monad m] (get : σ → σ')
+    (set : σ' → σ → σ) (x : state_t σ' m α) : state_t σ m α :=
+  ⟨λ st, do (a, st') ← x.run (get st),
+            pure (a, set st' st)⟩
```

Comment:
The state is not being used linearly here. I think this is an anti-idiom.
Suppose we use `zoom` for zooming into an array stored in the state. 
Then, when executing `x.run` we will have two references to the array, and destructive updates will not be performed.
Do you see a way to fix this? 
Do we really need `zoom`? AFAICT, all `monad_functor` machinery is only used to define `zoom` in this PR.

Reply:
I believe that in more complex monad-based code functions like `zoom` or `with_reader` for transforming monad stacks will be essential for gluing together different parts of the program. We can provide a specialized, linear version taking `(get : σ → σ') (set : σ' → σ)`, but the general `zoom` will be needed much more often I think. Usually you carry around some big state and want to give just a part of it to a subroutine.

Reply:
There may also be even more exotic signatures like `(get : σ → σ' → σ) (set : σ' → σ → σ)` where `get` is supposed to return a new σ with a "hole" where σ' was extracted. I wonder if that is a thing, like "linear lenses"?

Reply:
> We can provide a specialized, linear version taking (get : σ → σ') (set : σ' → σ)
This version is great for the zoom-out case. You have one example in the test folder where this signature would work well.

Reply:
@Kha I think you could accomodate that use case and manipulate the state linearly by combining the getter and setter into a lens: `lens : Π f [functor f], (σ' → f σ') → (σ → f σ)`. If that lens uses the state linearly -- I'm not sure how to make one that is computable but isn't linear -- the whole thing should be linear. And that lens is as expressive as a pair of getter and setter.

Reply:
> There may also be even more exotic signatures like (get : σ → σ' → σ) (set : σ' → σ → σ)

It should be `get : σ → (σ' x σ)`, right? 
If yes, then I can see it working. For the array example, `get` would retrieve the array from the state and return the state `st : σ` updated with an empty array. Thus, we would have a single reference to the array. Then, `set` would update the state returned by `get` with the array produced by `x.run`.
We could also have `(get : σ → (σ' x σ'')) (set : σ' → σ'' → σ)`, right?

Reply:
@cipher1024 I'm afraid my optics are not up to par to imagine what that would look like. Anyway, this may not be the best PR to introduce a lense framework :) .

Reply:
> It should be get : σ → (σ' x σ), right?

Reply:
> Yes. So, which one(s) to choose? :)

I think we can derive all variants from `(get : σ → (σ' x σ'')) (set : σ' → σ'' → σ)`, right?